test/e2e/firewall.go

```diff
@@ -0,0 +1,410 @@
+		firewallTestSourceRanges := []string{"0.0.0.0/1", "128.0.0.0/1"}
+
+		// OnlyLocal service is needed to examine which exact nodes the requests are being forward to by the Load Balancer on GCE
+		By("Creating a LoadBalancer type service with onlyLocal annotation")
+		serviceName := "firewall-test-loadbalancer"
+		jig := NewServiceTestJig(c, serviceName)
+		svc := createOnlyLocalLoadBalancServiceWithPort(jig, ns, serviceName, firewallTestHttpPort, firewallTestSourceRanges)
+		defer func() {
+			Expect(c.Core().Services(svc.Namespace).Delete(svc.Name, nil)).NotTo(HaveOccurred())
+		}()
+		svcExternalIP := svc.Status.LoadBalancer.Ingress[0].IP
+
+		By("Checking if service's firewall rules are correct")
+		Expect(examineSvcFirewallRules(gceCloud, svc)).NotTo(HaveOccurred())
+
+		By("Creating a daemon set of netexec pods ")
+		dsName := "firewall-test-daemon-set"
+		Expect(createNetexecDaemonSet(c, dsName, ns, jig.Labels, firewallTestHttpPort, firewallTestUdpPort, true)).NotTo(HaveOccurred())
+		defer func() {
+			deleteDaemonSet(f, dsName, ns, jig.Labels)
+		}()
+
+		By("Checking that daemon pods launch on every node of the cluster.")
+		Expect(wait.Poll(dsRetryPeriod, dsRetryTimeout, checkRunningOnAllNodes(f, jig.Labels))).NotTo(HaveOccurred(),
+			"error waiting for daemon pod to start")
+		Expect(checkDaemonStatus(f, dsName)).NotTo(HaveOccurred())
+
+		// Send requests from outside of the cluster because internal traffic is whitelisted
+		By("Accessing the external service ip from outside, all non-master nodes should be reached")
+		nodeNames, err := getNonMaterNodeNames(c)
+		Expect(err).NotTo(HaveOccurred())
+		Expect(examineHitNodesFromOutside(svcExternalIP, firewallTestHttpPort, firewallTimeoutDefault, nodeNames, true)).NotTo(HaveOccurred())
+
+		By("Removing tags from one of the nodes")
+		nodeNamesList := nodeNames.UnsortedList()
+		if len(nodeNamesList) <= 0 {
+			framework.Failf("Expect %v have at least 1 node", nodeNamesList)
+		}
+		removedTags, err := setNodeTags(gceCloud, cloudConfig.ProjectID, cloudConfig.Zone, nodeNamesList[0], []string{})
+		Expect(err).NotTo(HaveOccurred())
+		defer func() {
+			By("Adding tags back to the node")
+			_, err := setNodeTags(gceCloud, cloudConfig.ProjectID, cloudConfig.Zone, nodeNamesList[0], removedTags)
+			Expect(err).NotTo(HaveOccurred())
+			// Make sure the traffic is recover before exit
+			Expect(examineHitNodesFromOutside(svcExternalIP, firewallTestHttpPort, firewallTimeoutDefault, nodeNames, true)).NotTo(HaveOccurred())
+		}()
+
+		By("Accessing serivce through the external ip and examine got no response from the node without tags")
+		Expect(examineNotHitNodeFromOutside(svcExternalIP, firewallTestHttpPort, firewallTimeoutDefault, nodeNamesList[0], 15)).NotTo(HaveOccurred())
+	})
+})
+
+func createOnlyLocalLoadBalancServiceWithPort(jig *ServiceTestJig, ns, serviceName string, port int32, sourceRanges []string) *api.Service {
+	svc := jig.CreateTCPServiceOrFail(ns, func(svc *api.Service) {
```

Comment:
can you reuse something from the services test by moving it into framework/services_utils or someplace? less code to maintain/update/debug when something changes globally.